### ISSUE TITLE
On `Forc.toml` change, `forc` should try to update only package nodes that appear in the `lock::Diff`

### ISSUE THREAD

**Issue by mitchmindtree** [FuelLabs/sway#900]
## Problem

Currently, `forc`'s approach to updating the lock file in response to changes in the manifest is pretty simple. It:

1. Loads the `BuildPlan` from the `Forc.lock`.
2. Validates the `Lock` by retrieving all dependencies of the project node from the graph into a set, and compares them to the set of dependencies loaded by the manifest.
3. If the sets don't match, recreate the `Lock` from scratch.

Step 3 involves fetching each package's source in order to determine the current head for the specified git branch or tag. While this is simple in implementation and should always result in a correct package graph, this can result in a couple issues:

- We end up doing more fetching than is necessary if only one or a subset of dependencies were changed.
- We may unintentionally update the pinned HEAD of deps unrelated to the dep that was added/removed/edited in the manifest.

## Solution

To solve this, we should improve the strategy for handling detected differences between the `Forc.lock` and `Forc.toml`. Steps 1 and 2 should remain the same, but the following should be something closer to:

3. If the sets don't match, determine the packages that have been removed, added or edited.
4. Handle each change accordingly:
    - For removed packages, remove the node and all its dependencies that aren't used elsewhere in the graph. We can do so by removing nodes in breadth-first-search order to make sure parents are always removed before children. If a visited node still has an edge to a parent, then it means it has a dependent elsewhere in the graph and should be skipped.
    - For added packages, `fetch_deps` only for that sub-graph and merge it into the main graph.
    - For edited packages, perform a "remove" of the old node and an "add" as above.

## Offline

The changes above should also allow us to achieve `offline` updates to the lock file in the case that packages have only been *removed*. However, we'll still require fetching repositories for package edits or additions. After landing the above changes, we could open another issue around potentially building a local index of known dependencies so that, in the case that `--offline` is specified and we need to edit or add nodes to a lock file, we can lookup our local index in a best-effort approach that doesn't require fetching from remotes.

**Comment by kayagokalp** [comment]
Thanks for the great solution proposal that seems perfect. I am currently implementing and would like to summarize my current status and approach so that we can be on the same page. 

First of all, I started with finding `added` and `removed` packages. I am doing this by finding the difference between the `BTreeSet` coming from the manifest file and the `BTreeSet` coming from the "current" `BuildPlan` that is created from the current lock file. For the edited packages I am not doing something special as the approach we are going to be taking is simply removing first and adding later and finding differences on these two `BTreeSet` would resolve to the old version of the dep as a removed dependency and new version as a newly added dependency.  

So If I am not mistaken, first handling the removed dependencies and then handling the newly added ones seems like also handling the edited ones correctly. 

My current approach:

1. Find added and removed packages between manifest and current lock file
2. Iterate over compilation order and search the current node's corresponding package in removed packages
  - if the current dependency is removed, do a BFS from that node. Iterate over visited nodes and if a visited node has only **1 edge** remove that node, otherwise skip.
  - if the current dependency is not removed, continue with the next dependency in compilation order.

I haven't looked into adding dependencies yet but thinking about how I am going to implement merging the subgraph part at the back of my head.

**Comment by kayagokalp** [comment]
So after working for a couple of hours I think I was on the wrong path and I am summarizing my *updated* current approach:

Do a BFS from the root node (the current project itself) and in the order that nodes are visited check if the current visited node is a removed dependency or it does not have an edge to any other node (since the parents are going to be removed first the dependencies of the parents are going to be left with 0 edges eventually) remove the node. 

This way, once I remove `std` from a dummy project:

1. Remove `std` since it is going to be in the `removed dependency` list
2. Since std is removed `core` is left with no edges so it will be removed

**Comment by mitchmindtree** [comment]
@kayagokalp sounds like you're on the right track to me!

> check if the current visited node is a removed dependency or if it does not have an edge to any other node

I think this is what you meant, but just to clarify to be safe, this should be "does not have any **incoming** edge from any other node".

An edge ***a -> b*** implies that ***a*** depends on ***b***. If ***b*** only has one incoming edge and has been removed from the manifest, we can safely remove it from the graph as we know that no other packages depend on it.

However, if ***b*** contains more than one *incoming* edge, then there must be some other package in the graph that still depends on ***b*** so we leave it in the graph and just remove the edge from our top-level package.

**Comment by kayagokalp** [comment]
Started to look into the "adding" side of the things, directly connecting the root of the subgraph (that is going to be generated from the added packages) to the root of the current graph seems like a suitable solution but creates the following problem.

Imagine the following graph is created from the lock file.

<img width="382" alt="Screen Shot 2022-05-25 at 13 23 41" src="https://user-images.githubusercontent.com/20915464/170263030-8e107f01-5129-4c66-8884-2db7863e99f6.png">

And a dependency *e* is introduced in the manifest file so there is a mismatch between the lock and manifest file. 

The following dependency graph can be generated for dependency *e*,

<img width="262" alt="Screen Shot 2022-05-25 at 13 24 19" src="https://user-images.githubusercontent.com/20915464/170263817-32cf1457-b1eb-4419-ad9c-b626f52ef036.png">

Directly connecting the root of the subgraph for *e*, to the root of the initial dependency graph would yield something like,

<img width="346" alt="Screen Shot 2022-05-25 at 14 56 56" src="https://user-images.githubusercontent.com/20915464/170264061-f6cf89d1-2ee3-489c-8bc3-403186752487.png">

As it can be seen the same dependency *a* is present as 2 different nodes in the graph. I am planning to add subgraph nodes into the initial graph while pinning the dependencies so having two different *a* nodes for the same dependency *a* would result in *a* getting pinned two times. 

More ideal graph would be the following:

<img width="328" alt="Screen Shot 2022-05-25 at 16 23 27" src="https://user-images.githubusercontent.com/20915464/170272538-a6acc8d3-78bc-445a-9423-4833debc3057.png">

---
## Proposed Solution


Throw a HashMap into the problem 😎 , to have a mapping between dependencies to NodeIndexes. 

While inserting subgraph dependencies into the main graph (before pinning the new dependency) I will check if the node that I am going to be adding exists in the current graph, if so won't pin it, and will use the existing pinned dependency. Add the required edges to the existing node in the current graph rather than the duplicate of the node in the subgraph.

For testing out this approach, I created the following code:

```
use petgraph::dot::Dot;
use petgraph::graph::{Graph, NodeIndex};
use petgraph::visit::EdgeRef;
use std::collections::hash_map::Entry;
use std::collections::HashMap;
fn main() {
    let mut new_node_map: HashMap<NodeIndex, NodeIndex> = HashMap::new();
    let mut graph = Graph::<&str, &str>::new(); // directed and unlabeled
    let origin = graph.add_node("project_node");
    let a_node = graph.add_node("a");
    let b_node = graph.add_node("b");
    let c_node = graph.add_node("c");
    let d_node = graph.add_node("d");

    graph.add_edge(origin, a_node, "depends");
    graph.add_edge(origin, b_node, "depends");

    graph.add_edge(a_node, c_node, "depends");
    graph.add_edge(a_node, d_node, "depends");
    let mut graph2 = Graph::<&str, &str>::new(); // directed and unlabeled

    let e_node = graph2.add_node("e");
    let f_node = graph2.add_node("f");
    let a_node_2 = graph2.add_node("a");

    graph2.add_edge(e_node, f_node, "depends");
    graph2.add_edge(e_node, a_node_2, "depends");

    println!("{}", Dot::new(&graph));
    println!("{}", Dot::new(&graph2));
    let mut existing_node_map: HashMap<&str, NodeIndex> = HashMap::new();
    for node in graph.node_indices() {
        existing_node_map.insert(graph[node], node);
    }
    for node in graph2.node_indices() {
        if let Entry::Vacant(_) = existing_node_map.entry(graph2[node]) {
            let new_index = graph.add_node(graph2[node]);
            existing_node_map.insert(graph2[node], new_index);
            new_node_map.insert(node, new_index);
        } else {
            new_node_map.insert(node, existing_node_map[graph2[node]]);
        }
    }

    for edge in graph2.edge_references() {
        let new_p_index = new_node_map.get(&edge.source()).unwrap();
        let new_c_index = new_node_map.get(&edge.target()).unwrap();
        graph.add_edge(*new_p_index, *new_c_index, edge.weight());
    }

    graph.add_edge(origin, new_node_map[&e_node], "depends");
    println!("{}", Dot::new(&graph));
    println!("{}", Dot::new(&graph2));
}
```

If there are no edge cases that I am missing out with this approach, the only thing left is to move this logic into our code-base (with pinning stuff), and open the PR.

**Comment by mitchmindtree** [comment]
Nice, looking good!

One thing that might simplify this further would be to introduce the HashMap a little earlier while adding the nodes, e.g. you could do something like:

```rust
    // A map from the node source to its node index.
    let mut to_index: HashMap<&str, NodeIndex> = HashMap::new();
    // Adding a node to the graph could look like this:
    let a_node = *to_index.entry("a").or_insert_with(|| graph.add_node("a"));
```

Doing something along these lines would help to ensure that if a node already exists in the graph for the package, we always use the existing node's index first. A new node is only added in the case that there is no pre-existing node for the package.

**Comment by kayagokalp** [comment]
The transition of this logic into our codebase is done 🎉 , realized that we have this kind of logic happening with the `visited` hash map inside `fetch_children`. So I tried to re-use that part of the code. Currently, it is messy and I will be refactoring to something manageable and then will open a draft PR and do the rest of the refactoring there.

I tested the following scenario:

- Create a library named `child` which depends on `std` so we have something like `child` -> `std` -> `core`
- Create a project name `parent` which depends on `child`.

To test this I did:

1. forc init --library child
2. forc init parent
3. forc build in the parent directory to build it before depending on the child. 
4. add the child as a dependency for the parent
5. forc build in the parent directory

After step 3, this was the status of the dependency graph (sorry for the ultra-wide graph, for better visibility click and see in full-screen):

![after_step_2](https://user-images.githubusercontent.com/20915464/170502844-6de2a28d-a0d2-40fd-9332-ed2c78541bd2.png)

After step 5, the dependency graph became the following:

![after_step_5](https://user-images.githubusercontent.com/20915464/170503228-fe7eb891-4099-4f71-a2db-50519ed69175.png)

I observed the calls to the `fetch_children` by logging the name of the package that is being fetched by that function. The output in step 3 was the following:

```
fetch_children fetching parent_new
fetch_children fetching std
fetch_children fetching core
```

The output in step 5:

```
fetch_children fetching child_new
```

So only the newly added dependency (child) is fetched and while traversing the dependencies of that one (`std` -> `core`), a check from the `visited` hash map prevents us from re-fetching them.

---
### To compare with the current implementation

Following the testing steps with the current master,

The output in step 3:

```
fetch_children fetching parent_new
fetch_children fetching std
fetch_children fetching core
```

The output in step 5:
```
fetch_children fetching parent_new
fetch_children fetching child_new
fetch_children fetching std
fetch_children fetching core
```
